gfx_direct3d11.cpp

```diff
@@ -0,0 +1,976 @@
+    depth_stencil_texture_desc.ArraySize = 1;
+    depth_stencil_texture_desc.Format = DXGI_FORMAT_D32_FLOAT;
+    depth_stencil_texture_desc.SampleDesc = d3d.sample_description;
+    depth_stencil_texture_desc.Usage = D3D11_USAGE_DEFAULT;
+    depth_stencil_texture_desc.BindFlags = D3D11_BIND_DEPTH_STENCIL;
+    depth_stencil_texture_desc.CPUAccessFlags = 0;
+    depth_stencil_texture_desc.MiscFlags = 0;
+
+    ComPtr<ID3D11Texture2D> depth_stencil_texture;
+    ThrowIfFailed(d3d.device->CreateTexture2D(&depth_stencil_texture_desc, NULL, depth_stencil_texture.GetAddressOf()));
+    ThrowIfFailed(d3d.device->CreateDepthStencilView(depth_stencil_texture.Get(), NULL, d3d.depth_stencil_view.GetAddressOf()));
+
+    // Set render targets
+
+    d3d.context->OMSetRenderTargets(1, d3d.backbuffer_view.GetAddressOf(), d3d.depth_stencil_view.Get());
+
+    // Save resolution
+
+    d3d.current_width = width;
+    d3d.current_height = height;
+}
+
+LRESULT CALLBACK gfx_d3d11_dxgi_wnd_proc(HWND h_wnd, UINT message, WPARAM w_param, LPARAM l_param) {
+    switch (message) {
+        case WM_SIZE: {
+            RECT rect;
+            GetClientRect(h_wnd, &rect);
+            create_render_target_views(rect.right - rect.left, rect.bottom - rect.top);
+            break;
+        }
+        case WM_GETMINMAXINFO: {
+            LPMINMAXINFO lpMMI = (LPMINMAXINFO) l_param;
+            lpMMI->ptMinTrackSize.x = 320;
+            lpMMI->ptMinTrackSize.y = 240;
+            break;
+        }
+        case WM_DESTROY: {
+#if DEBUG_D3D
+            d3d.debug->ReportLiveDeviceObjects(D3D11_RLDO_DETAIL);
+#endif
+            exit(0);
+        }
+        case WM_ACTIVATEAPP: {
+            keyboard_on_all_keys_up();
+            break;
+        }
+        case WM_KEYDOWN: {
+            keyboard_on_key_down((l_param >> 16) & 0x1ff);
+            break;
+        }
+        case WM_KEYUP: {
+            keyboard_on_key_up((l_param >> 16) & 0x1ff);
+            break;
+        }
+        default: {
+            return DefWindowProcW(h_wnd, message, w_param, l_param);
+        }
+    }
+    return 0;
+}
+
+static void gfx_d3d11_dxgi_init(void) {
+
+    // Create window
+    WNDCLASSEXW wcex;
+    ZeroMemory(&wcex, sizeof(WNDCLASSEX));
+
+    wcex.cbSize = sizeof(WNDCLASSEX);
+    wcex.style = CS_HREDRAW | CS_VREDRAW;
+    wcex.lpfnWndProc = gfx_d3d11_dxgi_wnd_proc;
+    wcex.cbClsExtra = 0;
+    wcex.cbWndExtra = 0;
+    wcex.hInstance = nullptr;
+    wcex.hIcon = nullptr;
+    wcex.hCursor = LoadCursor(nullptr, IDC_ARROW);
+    wcex.hbrBackground = (HBRUSH)(COLOR_WINDOW + 1);
+    wcex.lpszMenuName = nullptr;
+    wcex.lpszClassName = WINCLASS_NAME;
+    wcex.hIconSm = nullptr;
+
+    RegisterClassExW(&wcex);
+
+    RECT wr = { 0, 0, DESIRED_SCREEN_WIDTH, DESIRED_SCREEN_HEIGHT };
+    AdjustWindowRect(&wr, WS_OVERLAPPEDWINDOW, FALSE);
+
+    h_wnd = CreateWindowW(WINCLASS_NAME, L"Super Mario 64 PC-Port (D3D11)", WS_OVERLAPPEDWINDOW,
```

Comment:
Hmm the game name should be in a define or be passed as argument to the renderer.

Reply:
I actually meant this repository should not be specific to any particular game. I meant that the one using this library should need to specify the name somewhere.

Reply:
Whoops, yeah, completely agree.

Reply:
I've made that the game's name is passed as an argument in gfx_init. And each graphics API adds its name to the title bar. Would that be ok?

Reply:
That seems nice!
I however see some mixed use of tabs and spaces. Can you change so that spaces are used only?

Reply:
Done, along with a couple of small fixes of mistakes I found.